internal/pkg/source/source.go

```diff
@@ -91,88 +91,156 @@ func (s *Source) dfs(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBl
 		return
 	}
 
+	if instr, ok := n.(ssa.Instruction); ok {
+		instrIndex, ok := indexInBlock(instr)
+		if !ok {
+			return
+		}
+
+		// If the referrer is in a different block from the one we last visited,
+		// and it can't be reached from the block we are visiting, then stop visiting.
+		if lastBlockVisited != nil && instr.Block() != lastBlockVisited && !s.canReach(lastBlockVisited, instr.Block()) {
+			return
+		}
+
+		// If this call's index is lower than the highest seen so far in its block,
+		// then this call is "in the past". If this call is a referrer,
+		// then we would be propagating taint backwards in time, so stop traversing.
+		// (If the call is an operand, then it is being used as a value, so it does
+		// not matter when the call occurred.)
+		if _, ok := instr.(*ssa.Call); ok && instrIndex < maxInstrReached[instr.Block()] && isReferrer {
+			return
+		}
+	}
+
 	mirCopy := map[*ssa.BasicBlock]int{}
 	for m, i := range maxInstrReached {
 		mirCopy[m] = i
 	}
 
 	if instr, ok := n.(ssa.Instruction); ok {
```

Comment:
I'm doing a second `if` that has the same head as the previous one because I wanted to detangle two things:
1. should we keep traversing?
2. update the max instruction in block, if applicable
These should really be replaced by calls to dedicated functions. I'm happy to do it in this PR, but I could do it in a future PR.

Reply:
I went ahead and extracted the first if (as well as other related instructions) into its own function. I think this reads much better.